table/block_based_table_builder.cc

```diff
@@ -101,83 +101,123 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;  // fall back to no compression.
+      return LZ4_Compress(
+          compression_ctx,
+          GetCompressFormatForVersion(kLZ4Compression, format_version),
+          raw.data(),
+          raw.size(),
+          compressed_output);
     case kLZ4HCCompression:
-      if (LZ4HC_Compress(
-              compression_ctx,
-              GetCompressFormatForVersion(kLZ4HCCompression, format_version),
-              raw.data(), raw.size(), compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;     // fall back to no compression.
+      return LZ4HC_Compress(
+          compression_ctx,
+          GetCompressFormatForVersion(kLZ4HCCompression, format_version),
+          raw.data(),
+          raw.size(),
+          compressed_output);
     case kXpressCompression:
-      if (XPRESS_Compress(raw.data(), raw.size(),
-          compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;
+      return XPRESS_Compress(raw.data(), raw.size(), compressed_output);
     case kZSTD:
     case kZSTDNotFinalCompression:
-      if (ZSTD_Compress(compression_ctx, raw.data(), raw.size(),
-                        compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;     // fall back to no compression.
-    default: {}  // Do not recognize this compression type
+      return ZSTD_Compress(
+          compression_ctx, raw.data(), raw.size(), compressed_output);
+    default:
+      // Do not recognize this compression type
+      return false;
   }
+}
+
+} // namespace
 
-  // Compression method is not supported, or not good compression ratio, so just
-  // fall back to uncompressed form.
+// format_version is the block format as defined in include/rocksdb/table.h
+Slice CompressBlock(
+    const Slice& raw,
+    const CompressionContext& compression_ctx,
+    CompressionType* type,
+    uint32_t format_version,
+    std::string* compressed_output,
+    std::string* sampled_output_fast,
+    std::string* sampled_output_slow) {
+  *type = compression_ctx.type();
+
+  if (compression_ctx.type() == kNoCompression &&
+      !compression_ctx.sample_for_compression()) {
+    return raw;
+  }
+
+  // If requested, we sample one in every N block with a
+  // fast and slow compression algorithm and report the stats.
+  // The users can use these stats to decide if it is worthwhile
+  // enabling compression and they also get a hint about which
+  // compression algorithm wil be beneficial.
+  // FIXME-SD: replace random()
+  if (compression_ctx.sample_for_compression() &&
+      ((random() % compression_ctx.sample_for_compression()) == 0) &&
```

Comment:
Though calling % per a block is unlikely to generate noticeable performance regression, but I still feel like it's a good idea to restraint ourselves from using division freely in critical paths. Can you turn it into a multification instead?

Reply:
Will do.

Reply:
So I ended up using the interface OneIn(), which was an existing interface and already used int the write path. Turns out it also has a mod. It's highly unlikely that this has any impact given the context we are in. But this might be the right place to make an improvement if needed.